reps/2023-8-18-ray-on-spark-autoscaling.md

```diff
@@ -0,0 +1,275 @@
+        requirements that are defined in autoscaler config YAML file.
+        After node creation, generate a unique node_id and link the node id with
+        several node metadata like tags / labels.
+        """
+        ...
+    
+    def terminate_node(self, node_id):
+        """
+        Terminate node by provided node id.
+        """
+  
+    def external_ip(self, node_id):
+        """
+        Get node external ip. In Ray on spark NodeProvider implementation,
+        we need to use fake IP that equals to node id, see Key questions
+        "How to make `NodeProvider` backend support multiple Ray worker nodes
+        running on the same virtual machine"
+        """
+        ...
+
+    def internal_ip(self, node_id):
+        """
+        Get node internal ip. In Ray on spark NodeProvider implementation,
+        we need to use fake IP that equals to node id, see Key questions
+        "How to make `NodeProvider` backend support multiple Ray worker nodes
+        running on the same virtual machine"
+        """
+        ...
+
+
+    def is_running(self, node_id):
+        """
+        query node running status.
+        """
+        ...
+
+    def is_terminated(self, node_id):
+        """
+        query whether node is terminated
+        """
+        ...
+
+    def node_tags(self, node_id):
+        """
+        query node tags
+        """
+        ...
+
+    def set_node_tags(self, node_id, tags):
+        """
+        set node tags
+        """
+        ...
+
+    @staticmethod
+    def bootstrap_config(cluster_config):
+        """
+        Override original autoscaler configs here.
+        """
+        ...
+```
+
+### How to implement Ray on spark node provider backend ?
+
+For Ray worker node creation, each Ray worker node, we start a spark job with only one spark
+task to hold this Ray worker node. When we need to terminate this Ray worker node, we cancel
+the corresponding spark task so that the Ray worker node is terminated.
+
+The critical issue is, in the process that executing `NodeProvider` backends, how to get
+the spark session in parent process ? We can set up a spark job server that runs inside
+parent process with mulitple threaded mode. The overall architecture is:
+
+![ray on spark autoscaling architecture](https://github.com/ray-project/ray/assets/19235986/9f1f30bb-e395-4d98-8c8d-5fa60675bafa)
+
+
+### Key questions
+
+
+#### We can use `ray up` command to set up a ray cluster with autoscaling, why we don't call `ray up` command in ray on spark autoscaling implementation ?
+
+In ray on spark, we only provides a python API `setup_ray_cluster`, ray on spark does not have a CLI, so in `setup_ray_cluster` implementation, we need to generate autoscale config YAML file according to `setup_ray_cluster` argument values, and then launch the ray head node with "--autoscaling-config" option, in this way Ray on Spark code can manage Ray head node process and ray worker nodes easier.
```

Comment:
Another thing is that ray-on-spark autoscaling only supports a very limited subset of `cluster.yaml`: e.g. only a single worker type, no docker, no ssh so it's easier and less confusing to use a restricted API.